File: locrec_ros/locrec_ros/conversions.py
"""PointCloud2 and Odometry to numpy, with no ROS import anywhere in this file.

The node needs rclpy. This does not, which is the point: the conversion is the
part most likely to be wrong and the part hardest to test inside a ROS graph, so
it is written against the wire format rather than against the Python class. Any
object with the message's attributes works, including a plain namespace built in
a test.

Field layout is read from the message rather than assumed. A Velodyne driver, an
Ouster driver and a rosbag replay all describe x, y and z with the same names but
at different offsets and sometimes different datatypes, and a hard-coded stride is
the classic way to get a point cloud that looks plausible and is wrong.
"""
from __future__ import annotations

import numpy as np
from scipy.spatial.transform import Rotation

__all__ = ["DATATYPES", "odometry_to_pose", "pointcloud2_to_xyz", "stamp_to_ns"]

# sensor_msgs/PointField datatype constants, repeated here so this module does not
# import ROS. The numbers are part of the message definition and do not change.
DATATYPES = {
    1: np.dtype(np.int8),
    2: np.dtype(np.uint8),
    3: np.dtype(np.int16),
    4: np.dtype(np.uint16),
    5: np.dtype(np.int32),
    6: np.dtype(np.uint32),
    7: np.dtype(np.float32),
    8: np.dtype(np.float64),
}


def pointcloud2_to_xyz(msg, max_points: int | None = None) -> np.ndarray:
    """Return an (N, 3) float64 array of finite xyz points.

    ``msg`` needs ``height``, ``width``, ``point_step``, ``row_step``, ``fields``,
    ``is_bigendian`` and ``data``. Rows are unpacked with ``row_step`` rather than
    ``width * point_step`` because organised clouds pad their rows.

    Non-finite points are dropped. Most drivers mark an invalid return with NaN,
    and a NaN reaching the registration turns the whole Hessian into NaN, which
    surfaces as a detector that silently never fires.
    """
    names = {f.name: f for f in msg.fields}
    missing = {"x", "y", "z"} - set(names)
    if missing:
        raise ValueError(f"point cloud has no {sorted(missing)} field")

    raw = np.frombuffer(bytes(msg.data), dtype=np.uint8)
    n_rows = max(int(msg.height), 1)
    row_step = int(msg.row_step) or int(msg.width) * int(msg.point_step)
    expected = n_rows * row_step
    if raw.size < expected:
        raise ValueError(f"point cloud data is {raw.size} bytes, expected {expected}")
    raw = raw[:expected].reshape(n_rows, row_step)
    point_step = int(msg.point_step)
    usable = (row_step // point_step) * point_step
    raw = raw[:, :usable].reshape(-1, point_step)

    n_points = int(msg.width) * n_rows
    raw = raw[:n_points]

    out = np.empty((raw.shape[0], 3), dtype=np.float64)
    for i, axis in enumerate("xyz"):
        field = names[axis]
        dtype = DATATYPES[int(field.datatype)]
        dtype = dtype.newbyteorder(">" if msg.is_bigendian else "<")
        offset = int(field.offset)
        chunk = raw[:, offset : offset + dtype.itemsize]
        out[:, i] = np.ascontiguousarray(chunk).view(dtype).reshape(-1)

    out = out[np.isfinite(out).all(axis=1)]
    if max_points is not None and out.shape[0] > max_points:
        stride = int(np.ceil(out.shape[0] / max_points))
        out = out[::stride]
    return out


def stamp_to_ns(stamp) -> int:
    """A ``builtin_interfaces/Time`` as integer nanoseconds, so stamps compare exactly."""
    return int(stamp.sec) * 1_000_000_000 + int(stamp.nanosec)


def odometry_to_pose(msg) -> np.ndarray:
    """The pose in a ``nav_msgs/Odometry`` as a 4x4 matrix: child frame in odometry frame.

    ``msg`` needs ``pose.pose.position`` and ``pose.pose.orientation``. The quaternion
    is normalised. A zero or non-finite one is an error rather than an identity, for
    the same reason a missing point field is: an identity rotation is a plausible
    prior that is wrong.
    """
    p = msg.pose.pose.position
    q = msg.pose.pose.orientation
    t = np.array([p.x, p.y, p.z], dtype=float)
    quat = np.array([q.x, q.y, q.z, q.w], dtype=float)
    norm = float(np.linalg.norm(quat))
    if not np.isfinite(t).all() or not np.isfinite(norm) or norm < 1e-9:
        raise ValueError("odometry pose has a non-finite position or a zero quaternion")
    T = np.eye(4)
    T[:3, :3] = Rotation.from_quat(quat / norm).as_matrix()
    T[:3, 3] = t
    return T

File: locrec_estimator/src/conversions.cpp
#include "locrec_estimator/conversions.hpp"

#include <cmath>
#include <cstring>
#include <stdexcept>
#include <string>

namespace locrec_estimator
{

namespace
{

/// Read one field of one point, honouring its declared datatype and the cloud's endianness.
double readField(const uint8_t * point, uint32_t offset, uint8_t datatype, bool big_endian)
{
  const auto load = [&](std::size_t size) {
      uint8_t bytes[8] = {0};
      std::memcpy(bytes, point + offset, size);
      if (big_endian) {
        for (std::size_t i = 0; i < size / 2; ++i) {
          std::swap(bytes[i], bytes[size - 1 - i]);
        }
      }
      return *reinterpret_cast<uint64_t *>(bytes);
    };

  switch (datatype) {
    case sensor_msgs::msg::PointField::INT8: {
        int8_t v = 0;
        std::memcpy(&v, point + offset, 1);
        return v;
      }
    case sensor_msgs::msg::PointField::UINT8: {
        uint8_t v = 0;
        std::memcpy(&v, point + offset, 1);
        return v;
      }
    case sensor_msgs::msg::PointField::INT16: {
        const uint64_t raw = load(2);
        int16_t v = 0;
        std::memcpy(&v, &raw, 2);
        return v;
      }
    case sensor_msgs::msg::PointField::UINT16: {
        const uint64_t raw = load(2);
        uint16_t v = 0;
        std::memcpy(&v, &raw, 2);
        return v;
      }
    case sensor_msgs::msg::PointField::INT32: {
        const uint64_t raw = load(4);
        int32_t v = 0;
        std::memcpy(&v, &raw, 4);
        return v;
      }
    case sensor_msgs::msg::PointField::UINT32: {
        const uint64_t raw = load(4);
        uint32_t v = 0;
        std::memcpy(&v, &raw, 4);
        return v;
      }
    case sensor_msgs::msg::PointField::FLOAT32: {
        const uint64_t raw = load(4);
        float v = 0.0F;
        std::memcpy(&v, &raw, 4);
        return v;
      }
    case sensor_msgs::msg::PointField::FLOAT64: {
        const uint64_t raw = load(8);
        double v = 0.0;
        std::memcpy(&v, &raw, 8);
        return v;
      }
    default:
      throw std::invalid_argument(
              "unsupported PointField datatype " + std::to_string(static_cast<int>(datatype)));
  }
}

}  // namespace

locrec::Points pointCloud2ToXyz(
  const sensor_msgs::msg::PointCloud2 & msg, std::optional<int> max_points)
{
  const sensor_msgs::msg::PointField * fields[3] = {nullptr, nullptr, nullptr};
  const char * names[3] = {"x", "y", "z"};
  for (const auto & field : msg.fields) {
    for (int i = 0; i < 3; ++i) {
      if (field.name == names[i]) {
        fields[i] = &field;
      }
    }
  }
  for (int i = 0; i < 3; ++i) {
    if (fields[i] == nullptr) {
      throw std::invalid_argument(std::string("point cloud has no ") + names[i] + " field");
    }
  }

  const std::size_t n_rows = std::max<std::size_t>(msg.height, 1);
  const std::size_t point_step = msg.point_step;
  const std::size_t row_step = msg.row_step != 0 ? msg.row_step : msg.width * point_step;
  if (point_step == 0) {
    throw std::invalid_argument("point cloud has a zero point_step");
  }
  const std::size_t expected = n_rows * row_step;
  if (msg.data.size() < expected) {
    throw std::invalid_argument(
            "point cloud data is " + std::to_string(msg.data.size()) + " bytes, expected " +
            std::to_string(expected));
  }

  const std::size_t n_points = static_cast<std::size_t>(msg.width) * n_rows;
  const std::size_t per_row = row_step / point_step;

  locrec::Points out;
  out.reserve(std::min<std::size_t>(n_points, 1u << 20));
  for (std::size_t p = 0; p < n_points; ++p) {
    const std::size_t row = p / per_row;
    const std::size_t column = p % per_row;
    const uint8_t * point = msg.data.data() + row * row_step + column * point_step;
    const Eigen::Vector3d xyz(
      readField(point, fields[0]->offset, fields[0]->datatype, msg.is_bigendian),
      readField(point, fields[1]->offset, fields[1]->datatype, msg.is_bigendian),
      readField(point, fields[2]->offset, fields[2]->datatype, msg.is_bigendian));
    if (xyz.allFinite()) {
      out.push_back(xyz);
    }
  }

  if (max_points.has_value() && static_cast<int>(out.size()) > *max_points) {
    const std::size_t stride = static_cast<std::size_t>(
      std::ceil(static_cast<double>(out.size()) / *max_points));
    locrec::Points thinned;
    thinned.reserve(out.size() / stride + 1);
    for (std::size_t i = 0; i < out.size(); i += stride) {
      thinned.push_back(out[i]);
    }
    out = std::move(thinned);
  }
  return out;
}

int64_t stampToNs(const builtin_interfaces::msg::Time & stamp)
{
  return static_cast<int64_t>(stamp.sec) * 1000000000LL + static_cast<int64_t>(stamp.nanosec);
}

Eigen::Isometry3d odometryToPose(const nav_msgs::msg::Odometry & msg)
{
  const auto & p = msg.pose.pose.position;
  const auto & q = msg.pose.pose.orientation;
  const Eigen::Vector3d t(p.x, p.y, p.z);
  Eigen::Quaterniond quat(q.w, q.x, q.y, q.z);
  const double norm = quat.norm();
  if (!t.allFinite() || !std::isfinite(norm) || norm < 1e-9) {
    throw std::invalid_argument("odometry pose has a non-finite position or a zero quaternion");
  }
  quat.normalize();
  return locrec::makeT(quat.toRotationMatrix(), t);
}

}  // namespace locrec_estimator
